scanner/indexer/indexer.go

```diff
@@ -140,6 +146,19 @@ func (i *indexerImpl) IndexContainerImage(
 	return i.indexer.Index(ctx, manifest)
 }
 
+func getLayerHTTPClient(ctx context.Context, imgRef name.Reference, auth authn.Authenticator, timeout time.Duration) (*http.Client, error) {
+	reg := imgRef.Context().Registry
+	tr, err := transport.NewWithContext(ctx, reg, auth, http.DefaultTransport, nil)
+	if err != nil {
+		return nil, err
+	}
+	httpClient := http.Client{
+		Timeout:   timeout,
+		Transport: tr,
+	}
+	return &httpClient, nil
```

Comment:
I have spent the last few minutes looking for the article, but I can't find it. There was an article from a while back that mentioned performance concerns due to `httpClient := http.Client{}` allocating the variable on the stack and then returning a pointer to it makes the variable escape to the heap meanwhile doing `httpClient := &http.Client{}` from the start will already put the http.Client in the heap.

Long way of saying I think `httpClient := &http.Client{}` is preferred

Reply:
```
$ go run main.go
stack: 303.84µs
heap: 303.81µs
$ cat main.go
package main

import (
	"fmt"
	"time"
	"net/http"
)

func httpStack() *http.Client {
	httpClient := http.Client{}
	return &httpClient
}
		
func httpHeap() *http.Client {
	return &http.Client{}
}

func main() {
	const iterations = 1000000
	start := time.Now()
	for i := 0; i < iterations; i++ {
		_ = httpStack()
		
	}
	elapsed := time.Since(start)
	fmt.Printf("stack: %s\n", elapsed)
	start = time.Now()
	for i := 0; i < iterations; i++ {
		_ = httpHeap()
	}
	elapsed = time.Since(start)
	fmt.Printf("heap: %s\n", elapsed)
}
$
```

Reply:
I think this is neglectible and I will move forward as is.

Reply:
you're saying those .03µs mean nothing to you?? 😄 

It may be possible the compiler is smart enough to notice the values from the function calls are never used, so it is completely skipped?

---

Edit:

I found the article I was looking for: https://hmarr.com/blog/go-allocation-hunting/

From what I can tell, the issue in the article has to do more with the copying of the elements of the slice rather than the heap allocation. Something similar was found here: https://go-review.googlesource.com/c/go/+/355789

I modified your program to disallow function inlining to see if that made any difference. The heap allocations seems to consistently be faster, but just by about 6-7ms at most:

```
$ cat main.go
package main

import (
	"fmt"
	"time"
	"net/http"
)

//go:noinline
func httpStack() *http.Client {
	httpClient := http.Client{}
	return &httpClient
}
	
//go:noinline	
func httpHeap() *http.Client {
	return &http.Client{}
}

func main() {
	const iterations = 1000000
	start := time.Now()
	for i := 0; i < iterations; i++ {
		_ = httpStack()
	}
	elapsed := time.Since(start)
	fmt.Printf("stack: %s\n", elapsed)
	start = time.Now()
	for i := 0; i < iterations; i++ {
		_ = httpHeap()
	}
	elapsed = time.Since(start)
	fmt.Printf("heap: %s\n", elapsed)
}

$ go build -gcflags="-m -l" *.go 2>&1
# command-line-arguments
./main.go:11:2: moved to heap: httpClient
./main.go:17:9: &http.Client{} escapes to heap
./main.go:27:12: ... argument does not escape
./main.go:27:28: elapsed escapes to heap
./main.go:33:12: ... argument does not escape
./main.go:33:27: elapsed escapes to heap

$ ./main
stack: 42.795712ms
heap: 36.217335ms
```